packages/flutter/lib/src/cupertino/refresh.dart

```diff
@@ -0,0 +1,514 @@
+  /// no indicator UI will be shown but the [onRefresh] will still be invoked.
+  ///
+  /// [onRefresh] will be called when pulled far enough to trigger a refresh.
+  const CupertinoRefreshControl({
+    this.refreshTriggerPullDistance: _kDefaultRefreshTriggerPullDistance,
+    this.refreshIndicatorExtent: _kDefaultRefreshIndicatorExtent,
+    this.builder: buildSimpleRefreshIndicator,
+    this.onRefresh,
+  }) : assert(refreshTriggerPullDistance != null && refreshTriggerPullDistance > 0),
+       assert(refreshIndicatorExtent != null && refreshIndicatorExtent >= 0),
+       assert(
+         refreshTriggerPullDistance >= refreshIndicatorExtent,
+         'The refresh indicator cannot take more space in its final state '
+         'than the amount initially created by overscrolling'
+       );
+
+  /// The amount of overscroll the scrollable must be dragged to trigger a reload.
+  ///
+  /// Must not be null, must be larger than 0 and larger than [refreshIndicatorExtent].
+  ///
+  /// When overscrolled past this distance, [onRefresh] will be called if not
+  /// null and the [builder] will build in the [RefreshIndicatorMode.armed] state.
+  final double refreshTriggerPullDistance;
+
+  /// The amount of space the refresh indicator sliver will keep holding while
+  /// [onRefresh]'s [Future] is still running.
+  ///
+  /// Must not be null and must be positive. But can be 0, in which case the
+  /// sliver will start retracting back to 0 as soon as the refresh started.
+  ///
+  /// Must be smaller than [refreshTriggerPullDistance] since the sliver shouldn't
+  /// grow more after triggering the refresh.
+  final double refreshIndicatorExtent;
+
+  /// A builder that's called as this sliver's size changes and as the state
+  /// changes.
+  ///
+  /// A default simple Twitter-style pull-to-refresh indicator is provided if
+  /// not specified.
+  ///
+  /// Can be set to null, in which case nothing will be drawn in the overscrolled
+  /// space.
+  ///
+  /// Will not be called when the available space is 0 such as before any
+  /// overscroll.
+  final RefreshControlIndicatorBuilder builder;
+
+  /// Callback invoked when pulled by [refreshTriggerPullDistance].
+  ///
+  /// If provided, must return a [Future] which will keep the indicator in the
+  /// [RefreshIndicatorMode.refresh] state until the [Future] completes.
+  ///
+  /// Can be null in which case a single frame of [RefreshIndicatorMode.armed]
+  /// state will be drawn before going immediately to the [RefreshIndicatorMode.done]
+  /// where the sliver will start retracting.
+  final RefreshCallback onRefresh;
+
+  static const double _kDefaultRefreshTriggerPullDistance = 100.0;
+  static const double _kDefaultRefreshIndicatorExtent = 60.0;
+
+  /// Retrieve the current state of the CupertinoRefreshControl. The same as the
+  /// state that gets passed into the [builder] function. Used for testing.
+  @visibleForTesting
```

Comment:
prefer @protected

Reply:
i'd rather we didn't expose this. You can get the state in the test by getting the widget by string (use a predicate finder that examines the runtimeType.toString) then getting its state as dynamic (tester.state).

Reply:
ya, I was trying to figure out how to not expose this too. I can get to the state but the issue is getting the field in the state since the state class is private